service/history/ndc/history_replicator_test.go

```diff
@@ -357,3 +360,196 @@ func TestNewHistoryReplicator_newMutableState(t *testing.T) {
 	)
 	assert.NotNil(t, testReplicatorImpl.newMutableState(mockDomainCacheEntry, log.NewNoop()))
 }
+
+func TestApplyEvents_newReplicationTask_validateReplicateEventsRequest(t *testing.T) {
+	encodingType1 := types.EncodingTypeJSON
+	encodingType2 := types.EncodingTypeThriftRW
+
+	historyEvent1 := &types.HistoryEvent{
+		Version: 0,
+	}
+	historyEvent2 := &types.HistoryEvent{
+		Version: 1,
+	}
+
+	historyEvents := []*types.HistoryEvent{historyEvent1, historyEvent2}
+	serializer := persistence.NewPayloadSerializer()
+	serializedEvents, err := serializer.SerializeBatchEvents(historyEvents, common.EncodingTypeThriftRW)
+	assert.NoError(t, err)
+
+	historyEvent3 := &types.HistoryEvent{
+		ID:      0,
+		Version: 2,
+	}
+	historyEvent4 := &types.HistoryEvent{
+		ID:      1,
+		Version: 3,
+	}
+
+	historyEvents2 := []*types.HistoryEvent{historyEvent3}
+	serializedEvents2, err := serializer.SerializeBatchEvents(historyEvents2, common.EncodingTypeThriftRW)
+	assert.NoError(t, err)
+
+	historyEvents3 := []*types.HistoryEvent{historyEvent3, historyEvent4}
+	serializedEvents3, err := serializer.SerializeBatchEvents(historyEvents3, common.EncodingTypeThriftRW)
+	assert.NoError(t, err)
+
+	historyEvents4 := []*types.HistoryEvent{historyEvent4}
+	serializedEvents4, err := serializer.SerializeBatchEvents(historyEvents4, common.EncodingTypeThriftRW)
+	assert.NoError(t, err)
+
+	tests := map[string]struct {
+		request          *types.ReplicateEventsV2Request
+		eventsAffordance func() []*types.HistoryEvent
+		expectedErrorMsg string
+	}{
+		"Case1: empty case": {
```

Comment:
remark, no action required: I generally suggest starting from the full case first since it's the most laborious, and then working through the exception cases, it is just easier to delete stuff than to add it

Reply:
actually, are all of these jsut error cases? Is there a test somewhere where we verify a successful application?

Reply:
Not yet. There are so many branches in one function, I was trying to build a full case first, but looks like it is more efficient to build the successful case step by step from different fail cases.